lib/private/User/Database.php

```diff
@@ -368,46 +369,62 @@ public function checkPassword(string $loginName, string $password) {
 	/**
 	 * Load an user in the cache
 	 *
-	 * @param string $uid the username
+	 * @param string $loginName the username or email
 	 * @return boolean true if user was found, false otherwise
 	 */
-	private function loadUser($uid) {
-		$this->fixDI();
+	private function loadUser(string $loginName, bool $tryEmail = true): bool {
+		if (isset($this->cache[$loginName])) {
+			return true;
+		}
 
-		$uid = (string)$uid;
-		if (!isset($this->cache[$uid])) {
-			//guests $uid could be NULL or ''
-			if ($uid === '') {
-				$this->cache[$uid] = false;
-				return true;
-			}
+		//guests $uid could be NULL or ''
+		if ($loginName === '') {
+			$this->cache[$loginName] = false;
+			return true;
+		}
+
+		$dbConn = $this->getDbConnection();
+		$qb = $dbConn->getQueryBuilder();
+		$qb->select('uid', 'displayname', 'password')
+			->from($this->table)
+			->where(
+				$qb->expr()->eq(
+					'uid_lower', $qb->createNamedParameter(mb_strtolower($loginName))
+				)
+			);
+		$result = $qb->executeQuery();
+		$row = $result->fetch();
+		$result->closeCursor();
+
+		// "uid" is primary key, so there can only be a single result
+		if ($row !== false) {
+			$this->cache[$loginName] = [
+				'uid' => (string)$row['uid'],
+				'displayname' => (string)$row['displayname'],
+				'password' => (string)$row['password'],
+			];
+			return true;
+		}
 
-			$qb = $this->dbConn->getQueryBuilder();
-			$qb->select('uid', 'displayname', 'password')
-				->from($this->table)
-				->where(
-					$qb->expr()->eq(
-						'uid_lower', $qb->createNamedParameter(mb_strtolower($uid))
-					)
-				);
-			$result = $qb->executeQuery();
-			$row = $result->fetch();
-			$result->closeCursor();
-
-			// "uid" is primary key, so there can only be a single result
-			if ($row !== false) {
-				$this->cache[$uid] = [
-					'uid' => (string)$row['uid'],
-					'displayname' => (string)$row['displayname'],
-					'password' => (string)$row['password'],
-				];
-			} else {
-				$this->cache[$uid] = false;
-				return false;
+		// Not found by UID so we try also for email, load uid for email.
+		if ($tryEmail) {
+			@[$emailUId] = $this->config->getUsersForUserValue('settings', 'email', mb_strtolower($loginName));
```

Comment:
no `@` operator please. Especially on a list construction like this it’s highly confusing.

Reply:
>  This operator is controversial. It may hide useful errors messages from function calls. A lot of programmers will avoid it even for hight cost. For assignments it is safe though.

Thought would be ok here as it seems to be safe on assignment.
The only reason for it is to silence the PHP notice on potentially null value, but that is safe, as if there is no first key in the array then `emailUId` is simply `null`.

Reply:
So the alternative without `@` would be this?

```suggestion
			$emailUId = reset($this->config->getUsersForUserValue('settings', 'email', mb_strtolower($loginName)));
```

and then check for `$emailUId !== false` below

Reply:
Yes, reset sounds good.
Alternative is:
```suggestion
			[$emailUId] = [...$this->config->getUsersForUserValue('settings', 'email', mb_strtolower($loginName)), null];
```

Not pretty.

Reply:
case should be `emailUid`, not `emailUId`. But the name is confusing actually.
Can we simply use `$uid` here?

Reply:
> case should be

was thinking like: email **u**ser **i**d

>  Can we simply use $uid here?

Agree that's better!